Xamarin.Forms.Xaml.UnitTests/DesignTimeLoaderTests.cs

```diff
@@ -0,0 +1,265 @@
+
+			var xaml = @"
+				<ContentPage xmlns=""http://xamarin.com/schemas/2014/forms""
+					xmlns:local=""clr-namespace:MissingNamespace;assembly=MissingAssembly""
+					xmlns:x=""http://schemas.microsoft.com/winfx/2009/xaml"">
+					<ContentPage.Resources>
+						<Style x:Key=""Test"" TargetType=""local:MyCustomButton"">
+							<Setter Property=""BackgroundColor"" Value=""Red"" />
+						</Style>
+					</ContentPage.Resources>
+					<local:MyCustomButton Style=""{StaticResource Test}"" />
+				</ContentPage>";
+
+			var page = (ContentPage)XamlLoader.Create(xaml, true);
+			Assert.That(page.Content, Is.TypeOf<Button>());
+			Assert.That(page.Content.BackgroundColor, Is.EqualTo(new Color(1, 0, 0)));
+		}
+
+		[Test]
+		public void ImplicitStyleAppliedToMissingType()
+		{
+			XamlLoader.FallbackTypeResolver = (p, type) => type ?? typeof(Button);
+
+			var xaml = @"
+				<ContentPage xmlns=""http://xamarin.com/schemas/2014/forms""
+					xmlns:local=""clr-namespace:MissingNamespace;assembly=MissingAssembly"">
+					<ContentPage.Resources>
+						<Style TargetType=""local:MyCustomButton"">
+							<Setter Property=""BackgroundColor"" Value=""Red"" />
+						</Style>
+					</ContentPage.Resources>
+					<local:MyCustomButton />
+				</ContentPage>";
+
+			var page = (ContentPage)XamlLoader.Create(xaml, true);
+			Assert.That(page.Content, Is.TypeOf<Button>());
+			Assert.That(page.Content.BackgroundColor, Is.EqualTo(new Color(1, 0, 0)));
+		}
+
+		[Test]
+		public void StyleTargetingRealTypeNotAppliedToUnknownType()
+		{
+			XamlLoader.FallbackTypeResolver = (p, type) => type ?? typeof(Button);
+
+			var xaml = @"
+				<ContentPage xmlns=""http://xamarin.com/schemas/2014/forms""
+					xmlns:local=""clr-namespace:MissingNamespace;assembly=MissingAssembly"">
+					<ContentPage.Resources>
+						<Style TargetType=""Button"">
+							<Setter Property=""BackgroundColor"" Value=""Red"" />
+						</Style>
+					</ContentPage.Resources>
+					<local:MyCustomButton />
+				</ContentPage>";
+
+			var page = (ContentPage)XamlLoader.Create(xaml, true);
+			Assert.That(page.Content, Is.TypeOf<Button>());
+			//Button Style shouldn't apply to MyButton
+			Assert.That(page.Content.BackgroundColor, Is.Not.EqualTo(Color.Red));
+		}
+
+		[Test]
+		public void StyleTargetingUnknownTypeNotAppliedToFallbackType()
+		{
+			XamlLoader.FallbackTypeResolver = (p, type) => type ?? typeof(Button);
+
+			var xaml = @"
+				<ContentPage xmlns=""http://xamarin.com/schemas/2014/forms""
+					xmlns:local=""clr-namespace:MissingNamespace;assembly=MissingAssembly"">
+					<ContentPage.Resources>
+						<Style TargetType=""local:MyCustomButton"">
+							<Setter Property=""BackgroundColor"" Value=""Red"" />
+						</Style>
+					</ContentPage.Resources>
+					<Button />
+				</ContentPage>";
+
+			var page = (ContentPage)XamlLoader.Create(xaml, true);
+
+			var myButton = (VisualElement)page.Content;
+			myButton._mergedStyle.ReRegisterImplicitStyles("MissingNamespace.MyButton");
```

Comment:
If the style is *not* being applied then should this line not be here?

Reply:
Also, we do not want to have to walk the forms hierarchy to attach styles (or anything) to arbitrary elements at arbitrary depths in the tree. I would expect this to be handled as part of the parser. Can we chat about this so we can figure out if there is a better approach here?